numpy/core/src/umath/ufunc_object.c

```diff
@@ -2057,6 +2111,11 @@ PyUFunc_GeneralizedFunctionInternal(PyUFuncObject *ufunc,
 
     NPY_UF_DBG_PRINT1("\nEvaluating ufunc %s\n", ufunc_name);
 
+    if (validate_casting(ufuncimpl,
```

Comment:
I've slightly lost track, but it seems `validate_casting` is called in quite a few places. Are there no duplications? Could one assume it is OK here?

Though if I trust the coverage, then obviously this branch is triggered at least sometime!

Reply:
It is called once for normal ufuncs, and once for gufuncs.  The "odd" thing is the trivial-loop special case/preparation.  That does not use it, but falls back to the slow path when in doubt.